File: backend/app/schemas/reminder.py
import uuid
from datetime import date, datetime, time

from pydantic import BaseModel, Field

from app.schemas.common import ORMModel

RTYPE = r"^(watering|treatment|fertilizing|recheck|other)$"


class ReminderIn(BaseModel):
    crop_id: uuid.UUID | None = None
    diagnosis_id: uuid.UUID | None = None
    title: str = Field(min_length=1, max_length=150)
    description: str | None = None
    reminder_type: str | None = Field(default="other", pattern=RTYPE)
    reminder_date: date
    reminder_time: time | None = None


class ReminderUpdate(BaseModel):
    crop_id: uuid.UUID | None = None
    title: str | None = Field(default=None, min_length=1, max_length=150)
    description: str | None = None
    reminder_type: str | None = Field(default=None, pattern=RTYPE)
    reminder_date: date | None = None
    reminder_time: time | None = None
    is_completed: bool | None = None


class ReminderOut(ORMModel):
    id: uuid.UUID
    crop_id: uuid.UUID | None = None
    crop_name: str | None = None
    diagnosis_id: uuid.UUID | None = None
    title: str
    description: str | None = None
    reminder_type: str | None = None
    reminder_date: date
    reminder_time: time | None = None
    is_completed: bool
    created_at: datetime | None = None


class NotificationOut(ORMModel):
    id: uuid.UUID
    type: str | None = None
    title: str | None = None
    body: str | None = None
    reference_id: uuid.UUID | None = None
    is_read: bool
    created_at: datetime | None = None

File: mobile/lib/domain/entities/reminder.dart
import '../../core/utils/format.dart';

class Reminder {
  Reminder({
    required this.id,
    required this.title,
    required this.date,
    this.description,
    this.type,
    this.time,
    this.cropId,
    this.cropName,
    this.diagnosisId,
    this.isCompleted = false,
  });
  final String id, title;
  final String? description, type, time, cropId, cropName, diagnosisId;
  final DateTime date;
  final bool isCompleted;

  bool get isTreatment => type == 'treatment' || type == 'recheck';

  factory Reminder.fromJson(Map<String, dynamic> j) => Reminder(
        id: j['id'],
        title: j['title'],
        description: j['description'],
        type: j['reminder_type'],
        date: DateTime.parse(j['reminder_date']),
        time: j['reminder_time'],
        cropId: j['crop_id'],
        cropName: j['crop_name'],
        diagnosisId: j['diagnosis_id'],
        isCompleted: j['is_completed'] ?? false,
      );
}

class AppNotification {
  AppNotification({required this.id, this.type, this.title, this.body, this.referenceId, this.isRead = false, this.createdAt});
  final String id;
  final String? type, title, body, referenceId;
  final bool isRead;
  final DateTime? createdAt;

  factory AppNotification.fromJson(Map<String, dynamic> j) => AppNotification(
        id: j['id'],
        type: j['type'],
        title: j['title'],
        body: j['body'],
        referenceId: j['reference_id'],
        isRead: j['is_read'] ?? false,
        createdAt: parseDate(j['created_at']),
      );
}
